src/libraries/Common/src/System/Marvin.cs

```diff
@@ -24,49 +24,54 @@ public static int ComputeHash32(ReadOnlySpan<byte> data, ulong seed)
         /// <summary>
         /// Computes a 64-hash using the Marvin algorithm.
         /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static long ComputeHash(ReadOnlySpan<byte> data, ulong seed)
+            => ComputeHash(ref MemoryMarshal.GetReference(data), (uint)data.Length, p0: (uint)seed, p1: (uint)(seed >> 32));
+
+        private static unsafe long ComputeHash(ref byte rBuffer, nuint cbBuffer, uint p0, uint p1)
         {
-            uint p0 = (uint)seed;
-            uint p1 = (uint)(seed >> 32);
+            nuint currentOffset = 0;
 
-            if (data.Length >= sizeof(uint))
+            fixed (byte* pbBuffer = &rBuffer)
             {
-                ReadOnlySpan<uint> uData = MemoryMarshal.Cast<byte, uint>(data);
+                // Consume as many 4-byte chunks as possible.
 
-                for (int i = 0; i < uData.Length; i++)
+                if (cbBuffer >= 4)
                 {
-                    p0 += uData[i];
-                    Block(ref p0, ref p1);
+                    nuint stopOffset = cbBuffer & ~(nuint)3;
+                    do
+                    {
+                        p0 += Unsafe.ReadUnaligned<uint>(pbBuffer + currentOffset);
```

Comment:
Is `Unsafe.ReadUnaligned<T>` safer in terms of (un)alignment than `*(T*)ptr`?
So why it's written this way, and not like 
```suggestion
                        p0 += *(uint*)(pbBuffer + currentOffset);
```

(Although when compared for readability, I like the Unsafe.ReadUnaligned more, as it's more explicit).

Reply:
Yes, it's safer. Regardless of whether you're dereferencing a GC-tracked reference (type _T\&_) or a non-GC-tracked reference (type _T\*_), it's all just the `ldind_*` instruction under the covers. [Per docs](https://docs.microsoft.com/dotnet/api/system.reflection.emit.opcodes.ldind_i4), all such addresses are assumed to be aligned, and the `unaligned.` prefix is the way to signal to the JIT that this assumption might not hold.

TBH I'm not aware of any current architecture which forbids dereferencing unaligned `uint*` addresses, but such architecture might exist in the future, so it's good to knock these out early. It'd look even cleaner if we could declare the local type `UNALIGNED int32_t*` like in C++, then we could use the natural deref operator, but oh well. :)

Reply:
Thanks for the explanation 👍